sklearn/utils/validation.py

```diff
@@ -2610,3 +2611,283 @@ def _to_object_array(sequence):
+        estimator.n_features_in_ = n_features
+        return
+
+    if not hasattr(estimator, "n_features_in_"):
+        # Skip this check if the expected number of expected input features
+        # was not recorded by calling fit first. This is typically the case
+        # for stateless transformers.
+        return
+
+    if n_features != estimator.n_features_in_:
+        raise ValueError(
+            f"X has {n_features} features, but {estimator.__class__.__name__} "
+            f"is expecting {estimator.n_features_in_} features as input."
+        )
+
+
+def validate_data(
+    _estimator,
+    /,
+    X="no_validation",
+    y="no_validation",
+    reset=True,
+    validate_separately=False,
+    cast_to_ndarray=True,
+    **check_params,
+):
+    """Validate input data and set or check feature names and counts of the input.
+
+    This helper function should be used in an estimator that requires input
+    validation. This mutates the estimator and sets the `n_features_in_` and
+    `feature_names_in_` attributes if `reset=True`.
+
+    ..versionadded:: 1.6
+
+    Parameters
+    ----------
+    _estimator : estimator instance
+        The estimator to validate the input for.
+
+    X : {array-like, sparse matrix, dataframe} of shape \
+            (n_samples, n_features), default='no validation'
+        The input samples.
+        If `'no_validation'`, no validation is performed on `X`. This is
+        useful for meta-estimator which can delegate input validation to
+        their underlying estimator(s). In that case `y` must be passed and
+        the only accepted `check_params` are `multi_output` and
+        `y_numeric`.
+
+    y : array-like of shape (n_samples,), default='no_validation'
+        The targets.
+
+        - If `None`, `check_array` is called on `X`. If the estimator's
+          requires_y tag is True, then an error will be raised.
+        - If `'no_validation'`, `check_array` is called on `X` and the
+          estimator's requires_y tag is ignored. This is a default
+          placeholder and is never meant to be explicitly set. In that case
+          `X` must be passed.
+        - Otherwise, only `y` with `_check_y` or both `X` and `y` are
+          checked with either `check_array` or `check_X_y` depending on
+          `validate_separately`.
+
+    reset : bool, default=True
+        Whether to reset the `n_features_in_` attribute.
+        If False, the input will be checked for consistency with data
+        provided when reset was last True.
+        .. note::
+           It is recommended to call reset=True in `fit` and in the first
+           call to `partial_fit`. All other methods that validate `X`
+           should set `reset=False`.
+
+    validate_separately : False or tuple of dicts, default=False
+        Only used if y is not None.
+        If False, call validate_X_y(). Else, it must be a tuple of kwargs
+        to be used for calling check_array() on X and y respectively.
```

Comment:
```suggestion
        Only used if `y` is not `None`.
        If `False`, call `validate_X_y()`. Else, it must be a tuple of kwargs
        to be used for calling `check_array` on `X` and `y` respectively.
```

Reply:
Is validate_X_y even a function within sklearn anymore?  I can't find it anywhere.

Reply:
Sorry, it's `check_X_y`